nservicebus/upgrades/7to8/dependency-injection.md

```diff
@@ -0,0 +1,102 @@
+* [Castle](/nservicebus/dependency-injection/castlewindsor.md)
+* [StructureMap](/nservicebus/dependency-injection/structuremap.md)
+* [Ninject](/nservicebus/dependency-injection/ninject.md)
+* [Unity](/nservicebus/dependency-injection/unity.md)
+
+Instead of the container adapter packages, use the [externally managed container mode](/nservicebus/dependency-injection/#externally-managed-mode) to use a third party dependency injection container. See the [migrating to externally managed mode](#externally-managed-container-mode-migrating-to-externally-managed-mode) section for examples using common dependency injection containers.
+
+## Property injection
+
+Property injection is not covered by `Microsoft.Extensions.DependencyInjection.Abstractions`, therefore the NServiceBus default dependency injection container does not support property injection anymore. Property injection might be supported by third party containers that can be enabled using the [externally managed container mode](/nservicebus/dependency-injection/#externally-managed-mode).
+
+## UseContainer is deprecated
+
+The `UseContainer` API to integrate third party containers with NServiceBus has been removed as it does not align with the `Microsoft.Extensions.DependencyInjection.Abstractions` model. To use a custom dependency injection container with NServiceBus, use the [externally managed container mode](/nservicebus/dependency-injection/#externally-managed-mode).
+
+## Externally managed container mode
+
+The externally managed container mode allows to integrate third party dependency injection containers with NServiceBus that conform to the `Microsoft.Extensions.DependencyInjection.Abstractions` model via the `EndpointWithExternallyManagedContainer.Create` API. See the following sections for examples using common DI containers.
+
+### Migrating to externally managed mode
+
+#### Microsoft.Extensions.DependencyInjection
+
+The following snippet configures NServiceBus to use the Microsoft's default container implementation. This requires the `Microsoft.Extensions.DependencyInjection` NuGet package.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var endpoint = await startableEndpoint.Start(serviceCollection.BuildServiceProvider());
+```
+
+#### Autofac
+
+The following snippet configures NServiceBus to use Autofac as its dependency injection container. This requires the `Autofac.Extensions.DependencyInjection` NuGet package.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var containerBuilder = new ContainerBuilder();
+containerBuilder.Populate(serviceCollection);
+var autofacContainer = containerBuilder.Build();
+var endpointInstance = await startableEndpoint.Start(new AutofacServiceProvider(autofacContainer));
+```
```

Comment:
I'd configure Autofac first and then create the startable endpoint.

Reply:
are you suggesting moving `var containerBuilder = new ContainerBuilder();` up between line 54-55? The remaining order needs to remain I think?

Reply:
Which diminishes the importance. Scratch that. It's OK as it is now.